spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/DefaultCloudSqlJdbcInfoProvider.java

```diff
@@ -54,12 +58,49 @@ public String getJdbcUrl() {
             this.properties.getDatabaseName(),
             this.properties.getInstanceConnectionName());
 
+    // Build additional JDBC url parameters from the configuration.
+    Map<String, String> urlParams = new LinkedHashMap<>();
     if (StringUtils.hasText(properties.getIpTypes())) {
-      jdbcUrl += "&ipTypes=" + properties.getIpTypes();
+      urlParams.put("ipTypes", properties.getIpTypes());
     }
 
     if (properties.isEnableIamAuth()) {
-      jdbcUrl += "&enableIamAuth=true&sslmode=disable";
+      urlParams.put("enableIamAuth", "true");
+      urlParams.put("sslmode", "disable");
+    }
+    if (StringUtils.hasText(properties.getTargetPrincipal())) {
+      urlParams.put("cloudSqlTargetPrincipal", properties.getTargetPrincipal());
+    }
+    if (StringUtils.hasText(properties.getDelegates())) {
+      urlParams.put("cloudSqlDelegates", properties.getDelegates());
+    }
+    if (StringUtils.hasText(properties.getAdminRootUrl())) {
+      urlParams.put("cloudSqlAdminRootUrl", properties.getAdminRootUrl());
+    }
+    if (StringUtils.hasText(properties.getAdminServicePath())) {
+      urlParams.put("cloudSqlAdminServicePath", properties.getAdminServicePath());
+    }
+    if (StringUtils.hasText(properties.getAdminQuotaProject())) {
+      urlParams.put("cloudSqlAdminQuotaProject", properties.getAdminQuotaProject());
+    }
+    if (StringUtils.hasText(properties.getUniverseDomain())) {
+      urlParams.put("cloudSqlUniverseDomain", properties.getUniverseDomain());
```

Comment:
Does cloud-sql-jdbc-socket-factory also allow for customizing the full endpoint?

Reply:
Yes, you can fully replace the URL to the SqlAdmin api using the `cloudSqlAdminRootUrl` property and `cloudSqlAdminServicePath` properties.

Reply:
Got it, thank you! Could we also add documentation for these property to the `sql.adoc` if they're not already included?

Reply:
Done.